Add efficient primitives for str.strip().

advait-dixit opened the pull request:
Fixes mypyc/mypyc#1090.

Copying cpython implementation for strip, lstrip and rstrip to `str_ops.c`.

JukkaL commented on mypyc/test-data/run-strings.test:
```diff
@@ -774,3 +774,16 @@ def test_surrogate() -> None:
     assert repr("foobar\x00\xab\ud912\U00012345") == r"'foobar\x00«\ud912𒍅'"
+
+[case testStrip]
+# This is a negative test. strip variants without args does not use efficient primitives.
```
Is this comment out of date / incorrect?

advait-dixit replied:
It was out of date. Removed it.

JukkaL commented on mypyc/test-data/run-strings.test:
```diff
@@ -774,3 +774,16 @@ def test_surrogate() -> None:
+    s = "xxb2yy"
+    assert s.lstrip("xy") == "b2yy"
+    assert s.strip("xy") == "b2"
+    assert s.rstrip("xy") == "xxb2"
```
Can you test all string kinds and different character code ranges, such as these (and mixing these):
* Character codes between 128 and 255 (0x80 to 0xff)
* Character codes between 256 and 65535 (0x100 to 0xffff)
* Character codes 65536+ (0x10000+)

advait-dixit replied:
Done.

JukkaL commented on mypyc/lib-rt/str_ops.c:
```diff
@@ -5,6 +5,58 @@
 #include <Python.h>
 #include "CPy.h"
 
+// Copied from cpython.git:Objects/unicodeobject.c.
```
Please mention the Python version or commit date from where this is from.

advait-dixit replied:
Done.

JukkaL commented on mypyc/lib-rt/str_ops.c:
```diff
@@ -174,6 +226,116 @@ PyObject *CPyStr_RSplit(PyObject *str, PyObject *sep, CPyTagged max_split) {
+    BLOOM_MASK sepmask;
+    Py_ssize_t seplen;
+
+    kind = PyUnicode_KIND(self);
```
I think you need to call `PyUnicode_READY` on `self` and `sepobj` (and check the return values) on 3.9 at least, before you can call `PyUnicode_KIND` etc. You can check the relevant function in Python 3.9 to see how it's used.

advait-dixit replied:
Done.

JukkaL commented on mypyc/lib-rt/str_ops.c:
```diff
@@ -174,6 +226,116 @@ PyObject *CPyStr_RSplit(PyObject *str, PyObject *sep, CPyTagged max_split) {
+// Copied from do_strip function in cpython.git/Objects/unicodeobject.c.
+PyObject *_CPyStr_Strip(PyObject *self, int strip_type, PyObject *sep) {
+    if (sep == NULL || sep == Py_None) {
+        Py_ssize_t len, i, j;
```
Similar to above, I think you'll need to call `PyUnicode_READY`.

advait-dixit replied:
Done.

JukkaL reviewed (commented):
Thanks for the PR! Looks good overall, just a few things I noticed. Can you run some microbenchmarks to make sure the performance impact is as expected?

advait-dixit commented:
> Thanks for the PR! Looks good overall, just a few things I noticed. Can you run some microbenchmarks to make sure the performance impact is as expected?

I have pushed a new commit that addresses review comments.
Below is a micro-benchmark on on my laptop. It compares CPython vs mypyc@this branch vs mypyc@master.

```
(py313_venv) advait@advaits-mbp mypy % cat prof_strip.py 
def solve1(s: str) -> str:
    for i in range(200):
        s.strip()
    return "s"

(py313_venv) advait@advaits-mbp mypy % cat driver.py 
import prof_strip
import timeit
import random
random_characters = "".join([chr(random.randint(0x0020, 0x10FFFF)) for _ in range(20)])
print(timeit.timeit(lambda: prof_strip.solve1(random_characters), number=100000))

(py313_venv) advait@advaits-mbp mypy % python driver.py                            
0.38529958299477585

(py313_venv) advait@advaits-mbp mypy % mypyc prof_strip.py > /dev/null && python driver.py
0.22940995899261907

(py313_venv) advait@advaits-mbp mypy % rm -rf build *.so && git checkout master && mypyc prof_strip.py > /dev/null && python driver.py
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
0.45345445899874903

```

JukkaL reviewed (approved):
Thanks for updates! The performance improvement is great.
